abdm_integrator/hiu/views/consents.py

```diff
@@ -0,0 +1,225 @@
+                    error_message=HIUError.CUSTOM_ERRORS[HIUError.CODE_PATIENT_NOT_FOUND],
+                    detail_attr='patient.id'
+                )
+        # TODO Remove below exception handling once addressed on abha side
+        except requests.Timeout:
+            raise ABDMServiceUnavailable()
+        except requests.HTTPError as err:
+            error = ABDMRequestHelper.gateway_json_from_response(err.response)
+            raise ABDMGatewayError(error.get('code'), error.get('message'))
+
+    def gateway_consent_request_init(self, consent_data):
+        payload = ABDMRequestHelper.common_request_data()
+        payload['consent'] = consent_data
+        ABDMRequestHelper().gateway_post(HIUGatewayAPIPath.CONSENT_REQUEST_INIT, payload)
+        return payload['requestId']
+
+    def save_consent_request(self, gateway_request_id, consent_data, user):
+        consent_request = ConsentRequest(user=user, gateway_request_id=gateway_request_id, details=consent_data)
+        consent_request.update_user_amendable_details(consent_data['permission'], consent_data['hiTypes'])
+        return consent_request
+
+
+class GatewayConsentRequestOnInit(HIUGatewayBaseView):
+    def post(self, request, format=None):
+        GatewayConsentRequestOnInitSerializer(data=request.data).is_valid(raise_exception=True)
+        self.process_request(request.data)
+        return Response(status=HTTP_202_ACCEPTED)
+
+    def process_request(self, request_data):
+        consent_request = ConsentRequest.objects.get(
+            gateway_request_id=request_data['resp']['requestId']
+        )
+        if request_data.get('consentRequest'):
+            consent_request.consent_request_id = request_data['consentRequest']['id']
+            consent_request.status = ConsentStatus.REQUESTED
+        elif request_data.get('error'):
+            consent_request.status = ConsentStatus.ERROR
+            consent_request.error = request_data['error']
+        consent_request.save()
+
+
+class GatewayConsentRequestNotify(HIUGatewayBaseView):
+
+    def post(self, request, format=None):
+        GatewayConsentRequestNotifySerializer(data=request.data).is_valid(raise_exception=True)
+        # TODO Run as a background task in celery
+        GatewayConsentRequestNotifyProcessor(request.data).process_request()
+        return Response(status=HTTP_202_ACCEPTED)
+
+
+class GatewayConsentRequestNotifyProcessor:
+
+    def __init__(self, request_data):
+        self.request_data = request_data
+
+    @transaction.atomic
+    def process_request(self):
+        consent_status = self.request_data['notification']['status']
+        if consent_status == ConsentStatus.REVOKED:
+            return self.handle_revoked()
+        consent_request = ConsentRequest.objects.get(
+            consent_request_id=self.request_data['notification']['consentRequestId']
+        )
+        consent_request.update_status(consent_status)
```

Comment:
We should do this in a single query using https://docs.djangoproject.com/en/3.1/topics/db/queries/#topics-db-queries-update

Reply:
Good approach.
However I need `consent_request` for further processing and update won't return the instance.